pymc3/step_methods/smc_utils.py

```diff
@@ -78,3 +85,200 @@ def _posterior_to_trace(posterior, variables, model, var_info):
             size += new_size
         strace.record({k: v for k, v in zip(varnames, value)})
     return MultiTrace([strace])
+
+
+def metrop_kernel(
+    q_old,
+    old_tempered_logp,
+    proposal,
+    scaling,
+    accepted,
+    any_discrete,
+    all_discrete,
+    discrete,
+    n_steps,
+    prior_logp,
+    likelihood_logp,
+    beta,
+    ABC,
+):
+    """
+    Metropolis kernel
+    """
+    deltas = np.squeeze(proposal(n_steps) * scaling)
+    for n_step in range(n_steps):
+        delta = deltas[n_step]
+
+        if any_discrete:
+            if all_discrete:
+                delta = np.round(delta, 0).astype("int64")
+                q_old = q_old.astype("int64")
+                q_new = (q_old + delta).astype("int64")
+            else:
+                delta[discrete] = np.round(delta[discrete], 0)
+                q_new = floatX(q_old + delta)
+        else:
+            q_new = floatX(q_old + delta)
+
+        if ABC:
+            ll = likelihood_logp(q_new)
+        else:
+            ll = likelihood_logp(q_new)[0]
+
+        new_tempered_logp = prior_logp(q_new) + ll * beta
+
+        q_old, accept = metrop_select(new_tempered_logp - old_tempered_logp, q_new, q_old)
+        if accept:
+            accepted += 1
+            old_tempered_logp = new_tempered_logp
+
+    return q_old, accepted
+
+
+def calc_beta(beta, likelihoods, threshold=0.5):
+    """
+    Calculate next inverse temperature (beta) and importance weights based on current beta
+    and tempered likelihood.
+
+    Parameters
+    ----------
+    beta : float
+        tempering parameter of current stage
+    likelihoods : numpy array
+        likelihoods computed from the model
+    threshold : float
+        Determines the change of beta from stage to stage, i.e.indirectly the number of stages,
+        the higher the value of threshold the higher the number of stage. Defaults to 0.5.
+        It should be between 0 and 1.
+
+    Returns
+    -------
+    new_beta : float
+        tempering parameter of the next stage
+    old_beta : float
+        tempering parameter of the current stage
+    weights : numpy array
+        Importance weights (floats)
+    sj : float
+        Partial marginal likelihood
+    """
+    low_beta = old_beta = beta
+    up_beta = 2.0
+    rN = int(len(likelihoods) * threshold)
+
+    while up_beta - low_beta > 1e-6:
+        new_beta = (low_beta + up_beta) / 2.0
+        weights_un = np.exp((new_beta - old_beta) * (likelihoods - likelihoods.max()))
+        weights = weights_un / np.sum(weights_un)
+        ESS = int(1 / np.sum(weights ** 2))
+        if ESS == rN:
```

Comment:
Hmm, I am not sure I understand this part.

Reply:
The effective sample size can be estimated from the weights, empirical evidence suggest that good results are obtained when ESS = N/2. Here that observation is used to estimate the beta value to be used on the next stage.

Reply:
right, why `break` when ESS excatly equal to rN is what I not quite get.

Reply:
The ESS is indirectly measuring the uniformity of the weight. The ESS will be higher for a more uniform distribution of weights than for a peaked distribution. For importance sampling (and SMC) we want more or less uniform distribution of weights, otherwise we loose diversity (one or a few particles explains a lot and the other almost nothing). The ESS can range from 1 to the number of samples (N), with the default value of rN = N/2 being in the middle. If you move super slowly from stage to stage (change beta by a very small amount) then you will guarantee that the weights are very close uniform (and ESS close to N) but you will need to take many steps. If you move too fast (and the prior and posterior are very different) then you are at risk of a single particle having all (or most off the weights).  We use to use the covariance of the weights I change it to ESS to explore other definitions of ESS with the hope of finding an improved way to compute beta.